[API Shield] SV2 system (#14668)
cloudflare/cloudflare-docs: content/api-shield/security/schema-validation/_index.md

## Body inspection

API Shield has the ability to identify body specifications contained in uploaded schemas and validate that the data of incoming API requests adheres to them.

Schema validation currently supports validating requests with content-type `application/json`.

Within the OpenAPI specification, request body schemas are associated to media-ranges (such as `application/*`, `application/xml` or `application/json`).

> When Cloudflare validates incoming requests, Cloudflare checks that the request's content-type matches to the OpenAPI-specified media-range, for example:

Author: ```suggestion
When Cloudflare validates incoming requests, Cloudflare checks that the request's `content-type` matches the OpenAPI-specified media-range. 
```

> When the OpenAPI file specifies `application/*` as part of the request body content map, Cloudflare will accept requests with the content-types `application/xml` and `application/json`, etc. However, only `application/json` bodies will be validated with the supplied schema.

Author: ```suggestion
For example, when the OpenAPI file specifies `application/*` as part of the request body content map, Cloudflare will accept requests with the content-types `application/xml` and `application/json`, etc. However, only `application/json` bodies will be validated with the supplied schema.
```

> Note: We generally recommend keeping the media-ranges as tight as possible. If you need to support multiple content-types on an API endpoint, explicitly list those as individual media-ranges and refrain from using wildcard identifiers. Futhermore, care should be taken if the origin is configured to perform [MIME sniffing](https://mimesniff.spec.whatwg.org/). For example, when a request carrying a json body is deliberately carrying an `application/malicous` content-type and Cloudflare was configured to allow `application/*` media-ranges, the request would be passed along to the origin without validating the json body contents. However, an origin that ignores the content-type, and either trial deserializes or sniffs the MIME type may thus deserialize the json body with a wrong assumption about it having passed schema body validation.
> As such, if you need to support, for example, `application/json` and `application/xml` on the same endpoint, don't use `application/*` but rather set one content for `application/json` (where Cloudflare will check the request body with the supplied schema) and one with `application/xml` that Cloudflare will just let through. It is still advised to disable content-type sniffing on your origin.
> 
> Cloudflare allows specifying the following media-ranges in the OpenAPI request body content map: `*/*`, `application/*`, and `application/json`. Further, media-ranges can also be configured to enforce a `charset` parameter. For this, we only accept the `charset` parameter with a static value of `utf-8` as part of the media-range specification and, when configured, Cloudflare will similarly require the request's content-type to carry this charset.

Author: ```suggestion
Cloudflare generally recommends keeping the media-ranges as tight as possible. If you need to support multiple content-types on an API endpoint, explicitly list those as individual media-ranges and refrain from using wildcard identifiers. 

Care should also be taken if the origin is configured to perform [MIME sniffing](https://mimesniff.spec.whatwg.org/). For example, when a request carrying a JSON body is deliberately carrying an `application/malicous` content-type and Cloudflare was configured to allow `application/*` media-ranges, the request would be passed along to the origin without validating the JSON body contents. However, an origin that ignores the content-type, and either trial deserializes or sniffs the MIME type may deserialize the JSON body with a wrong assumption of having passed schema body validation.

As such, if you need to support `application/json` and `application/xml` on the same endpoint, do not use `application/*`. Instead, set one content for `application/json`, where Cloudflare will check the request body with the supplied schema, and another with `application/xml` that Cloudflare will let through. It is still advised to disable content-type sniffing on your origin.

Cloudflare allows specifying the following media-ranges in the OpenAPI request body content map: 
- `*/*`
- `application/*`
- `application/json`. 

Media-ranges can also be configured to enforce a `charset` parameter. For this, Cloudflare only accepts the `charset` parameter with a static value of `utf-8` as part of the media-range specification and when configured, we will similarly require the request's content-type to carry this charset.
```

## Body inspection

API Shield has the ability to identify body specifications contained in uploaded schemas and validate that the data of incoming API requests adheres to them.

Schema validation currently supports validating requests with content-type `application/json`.

Within the OpenAPI specification, request body schemas are associated to media-ranges (such as `application/*`, `application/xml` or `application/json`).

When Cloudflare validates incoming requests, Cloudflare checks that the request's `content-type` matches the OpenAPI-specified media-range. 
For example, when the OpenAPI file specifies `application/*` as part of the request body content map, Cloudflare will accept requests with the content-types `application/xml` and `application/json`, etc. However, only `application/json` bodies will be validated with the supplied schema.

> Cloudflare generally recommends keeping the media-ranges as tight as possible. If you need to support multiple content-types on an API endpoint, explicitly list those as individual media-ranges and refrain from using wildcard identifiers. 

Reviewer: ```suggestion
Cloudflare generally recommends keeping the media-ranges as tight as possible. We recommend, setting them to an individual media-type. If you need to support multiple content-types on an API endpoint, you can utilize wildcard media-ranges (see below).
```

Care should also be taken if the origin is configured to perform [MIME sniffing](https://mimesniff.spec.whatwg.org/). For example, when a request carrying a JSON body is deliberately carrying an `application/malicous` content-type and Cloudflare was configured to allow `application/*` media-ranges, the request would be passed along to the origin without validating the JSON body contents. However, an origin that ignores the content-type, and either trial deserializes or sniffs the MIME type may deserialize the JSON body with a wrong assumption of having passed schema body validation.

> As such, if you need to support `application/json` and `application/xml` on the same endpoint, do not use `application/*`. Instead, set one content for `application/json`, where Cloudflare will check the request body with the supplied schema, and another with `application/xml` that Cloudflare will let through. It is still advised to disable content-type sniffing on your origin.

Reviewer: ```suggestion
As such, if you need to support `application/json` and `application/xml` on the same endpoint, you can use `application/*`. Cloudflare will validate the provided schema for request bodies where the content-type is set to `application/json`. Requests with content-type `application/xml` (and others matching `application/*`) will be let through. It is still strongly advised to disable content-type sniffing on your origin.
```
